frontends/pytorch/csrc/c10_dispatch/acap_dispatch.cpp

```diff
@@ -34,9 +35,76 @@ using c10::Stack;
 // TODO: Ask the PT devs why conv is special and only shows up if dispatching
 // through the autograd keys.
 // https://github.com/llvm/mlir-npcomp/issues/86
-// #define ACAP_DISPATCH_KEY AutogradPrivateUse3
-#define ACAP_DISPATCH_KEY PrivateUse3
+#define ACAP_DISPATCH_KEY PrivateUse2
+#define ACAP_GRAD_DISPATCH_KEY AutogradPrivateUse2
 static c10::DispatchKey kAcapDispatchKey = c10::DispatchKey::ACAP_DISPATCH_KEY;
+static c10::DispatchKey kAcapGradDispatchKey =
+    c10::DispatchKey::ACAP_GRAD_DISPATCH_KEY;
+
+AcapController::KernelCallBuilder::KernelCallBuilder(AcapController &parent,
+                                                     MlirContext context,
+                                                     MlirLocation loc,
+                                                     std::string &kernelName)
+    : parent(parent), context(context), loc(loc), kernelName(kernelName),
+      state("torch.kernel_call", loc) {
+  (void)this->context; // Preserve for future.
+  MlirNamedAttribute kernelNameAttr = mlirNamedAttributeGet(
+      "kernel_name",
+      mlirStringAttrGet(context, kernelName.size(), kernelName.data()));
+  mlirOperationStateAddAttributes(state, 1, &kernelNameAttr);
+}
+
+void AcapController::KernelCallBuilder::addOperand(const IValue &value) {
+  MlirValue mlirValue = parent.mapIValueToMlirValue(loc, value);
+  if (mlirValueIsNull(mlirValue)) {
+    std::stringstream out;
+    out << "Unsupported capture value returned from kernel '" << kernelName
+        << "' (" << value.tagKind() << "): " << value;
+    throw std::invalid_argument(out.str());
+  }
+  mlirOperationStateAddOperands(state, 1, &mlirValue);
+}
+
+void AcapController::KernelCallBuilder::addResult(const IValue &value) {
+  MlirType resultType = parent.mapIValueToMlirType(loc, value);
+  if (mlirTypeIsNull(resultType)) {
+    std::stringstream out;
+    out << "Unsupported capture value returned from kernel '" << kernelName
+        << "' (" << value.tagKind() << "): " << value;
+    throw std::invalid_argument(out.str());
+  }
+  if (value.isTensor()) {
+    resultIndexToTensorMap.emplace_back(resultCount++, value.toTensor());
+  }
+  mlirOperationStateAddResults(state, 1, &resultType);
+}
+
+MlirOperation AcapController::KernelCallBuilder::create() {
+  // Create operation.
+  MlirOperation op = state.createOperation();
+  parent.funcBuilder->getEntryBlockBuilder().insertBeforeTerminator(op);
+
+  // Map result tensors.
+  for (auto &it : resultIndexToTensorMap) {
+    MlirValue result = mlirOperationGetResult(op, it.first);
+    parent.funcBuilder->mapTensor(it.second, result);
+  }
+
+  // Add to debug log.
+  std::stringstream sout;
+  sout << "CAPTURE: " << kernelName << "\n";
+  parent.captureLog.push_back(sout.str());
+  return op;
+}
+
+at::Tensor AcapController::importTensor(at::Tensor tensor) {
+  // TODO: Move the tensor to the CPU.
+  auto impl = c10::make_intrusive<at::TensorImpl>(
```

Comment:
seems a little strange for our code to be manipulating an "impl" object. Comment?